Scripts to create an Always On Availability Group in Google Cloud

sqlthinker opened the pull request:
The scripts will automatically create a SQL Server Availability Group

SurferJeffAtGoogle commented:
Thanks for the PR, Anibal.  It's going to take me awhile to review.  I'll probably reply with my first comments early next week.

SurferJeffAtGoogle commented on compute/sqlserver/powershell/create-availability-group.ps1:
```diff
@@ -0,0 +1,409 @@
+  Throw 'Need to install SQL Server in both nodes before running this script.'
+}
+
+# Only continue if script is ran from a computer that is member of the domain
```
Why?  Is it a security concern?  (Answer in comment.)

SurferJeffAtGoogle replied:
Ahh.  Ok.  Can you copy that into a comment?

SurferJeffAtGoogle commented on compute/sqlserver/powershell/create-availability-group.ps1:
```diff
@@ -0,0 +1,409 @@
+# Enable CredSSP in local computer
+Write-Host "$(Get-Date) Enable CredSSP Client in local machine"
+Enable-WSManCredSSP Client -DelegateComputer * -Force | Out-Null
+Start-Sleep 15
```
Why sleep?  What happens if the operation doen't complete in 15 minutes?

SurferJeffAtGoogle replied:
Ok.  Sounds like a reasonable solution.  Please say something about the random error in a comment.

SurferJeffAtGoogle commented on compute/sqlserver/powershell/create-sql-instance-availability-group-prerequisites.ps1:
```diff
@@ -0,0 +1,384 @@
+# Specify the parameters
+################################################################################
+# Network configuration
+$network      = 'wsfcnet'      # Name of the Custom network
```
It's not clear if users are expected to update these parameters with their own settings.

sqlthinker replied:
We have a document that goes step by step how to create a HA Availability Group. The document references the script create-sql-instance-availability-group.ps1 as a way to automate the steps. But, it also says that before running the script, they need to do some pre-requisites. We realized that some customers are not familiar with GCP and will not know how to do those pre-requisites. That is why we created this "-prerequisites" script for them. It matches the same values we used in our documentation. We expected that it will be used by clients following the instructions document and that will use exactly the same values we used in the doc.

SurferJeffAtGoogle replied:
If the doc is live, please add a link to it here.

SurferJeffAtGoogle commented on compute/sqlserver/powershell/create-sql-instance-availability-group-prerequisites.ps1:
```diff
@@ -0,0 +1,384 @@
+# Remove firewall rule to allow current IP
+#Get-GceFirewall -FirewallName "allow-winrm-current-ip" | Remove-GceFirewall
+
+##### Cleanup - Run with caution as everything will be deleted #####
```
Could we put the cleanup code into its own function or script?

sqlthinker replied:
Agree. I'll remove it from here and put it in a new script with the same name but with "-cleanup.ps1" at the end.

SurferJeffAtGoogle commented:
Like the bug fixes in f867696, please make changes in this PR as additional commits.

sqlthinker commented:
Added more comments based on feedback. Also added the -s to the calls to Start-Sleep for readability. Fixed some bugs where we checked the wrong network name when validating if the network already exists.

sqlthinker commented:
To test you can run the script "create-sql-instance-availability-group.Tests.ps1". It uses the Pester module to validate that the Availability Group was successfully created. But, be aware that the script needs to be run from a computer member of the domain. It is the only way to create the cluster because of the "double-hop" issue when creating the cluster using WinRM and CredSSP. You can modify the script "parameters-config.ps1" and specify the password for the account to add the nodes to the domain. The default value is "$null" which means the script will prompt the user for a password, which may be an issue if you are trying to automate it.
